### PR TITLE
client: Don't write permission when the file is marked with SUID or SGID

### PR DESCRIPTION
**Pull request by renhwztetecs**
… SUID or SGID

The problem appears in the ceph-fuse mount,there is no problem with the kernel mount.
we can add the identification of S_ISUID and S_ISGID in may_setattr.

Fixes: https://tracker.ceph.com/issues/45320
Signed-off-by: huanwen ren <ren.huanwen@zte.com.cn>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -5421,6 +5421,12 @@ int Client::may_setattr(Inode *in, struct ceph_statx *stx, int mask,
   int r = _getattr_for_perm(in, perms);
   if (r < 0)
     goto out;
+  
+  if (in->mode & S_ISUID)
+    goto out;
+
+  if ((in->mode & (S_ISGID | S_IXGRP)) == (S_ISGID | S_IXGRP))
+    goto out;
```
Comment (ukernel) [inline src/client/Client.cc:5429]: this allow all operation for file with setuid/setgid, which is incorrect.
In may_setattr CEPH_SETATTR_MODE backet, we should allow removing SUID/SGID if file write is allowed for uid.
A better fix is set FUSE_HANDLE_KILLPRIV fuse flag, and let libcephfs handle removing suid/sgid on write.

Comment (ukernel) [reply src/client/Client.cc:5429]: something like:
<pre>
diff --git a/src/client/Client.cc b/src/client/Client.cc
index 84fcf91d0d..25a917c704 100644
--- a/src/client/Client.cc
+++ b/src/client/Client.cc
@@ -5394,8 +5394,16 @@ int Client::may_setattr(Inode *in, struct ceph_statx *stx, int mask,
   }
 
   if (mask & CEPH_SETATTR_MODE) {
-    if (perms.uid() != 0 && perms.uid() != in->uid)
-      goto out;
+    if (perms.uid() != 0 && perms.uid() != in->uid) {
+      auto to_clear = ~stx->stx_mode & in->mode & 07777;
+      if (to_clear && !(to_clear & ~(S_ISUID | S_ISGID)) && // only remove setuid/setgid
+         !(stx->stx_mode & ~in->mode & 07777) && // not add new bits
+         !inode_permission(in, perms, MAY_WRITE)) { // may write
+       // hack: kill setuid/setguid on write?
+      } else {
+       goto out;
+      }
+    }
 
     gid_t i_gid = (mask & CEPH_SETATTR_GID) ? stx->stx_gid : in->gid;
     if (perms.uid() != 0 && !perms.gid_in_groups(i_gid))

</pre>

Comment (renhwztetecs) [reply src/client/Client.cc]: Thank you
FUSE_HANDLE_KILLPRIV fuse flag is the best way , but only supported in fuse3.0

Comment (renhwztetecs) [reply src/client/Client.cc]: ！I first try to test it

File: src/client/Client.cc [hunk 2]
```diff
@@ -5440,8 +5440,16 @@ int Client::may_setattr(Inode *in, struct ceph_statx *stx, int mask,
   }
 
   if (mask & CEPH_SETATTR_MODE) {
-    if (perms.uid() != 0 && perms.uid() != in->uid)
-      goto out;
+    if (perms.uid() != 0 && perms.uid() != in->uid) {
+      if(in->caps_issued_mask(CEPH_CAP_AUTH_EXCL)) {
```
Comment (ukernel) [inline src/client/Client.cc:5444]: why do you check CEPH_CAP_AUTH_EXCL? It's possible that multiple clients open a setuid file for write at the same time.

File: src/client/Client.cc (hunk 1 above)
Comment (renhwztetecs) [reply src/client/Client.cc]: The provided method cannot be verified, and the write function will clear out suid and sgid

File: src/client/Client.cc (hunk 2 above)
Comment (renhwztetecs) [reply src/client/Client.cc]: This place is wrong, I modified it, please review again.
I use the pjdfstest  tool for a test, the compatibility result is correct

File: src/client/Client.cc [hunk 3]
```diff
@@ -5440,8 +5440,16 @@ int Client::may_setattr(Inode *in, struct ceph_statx *stx, int mask,
   }
 
   if (mask & CEPH_SETATTR_MODE) {
-    if (perms.uid() != 0 && perms.uid() != in->uid)
-      goto out;
+    if (perms.uid() != 0 && perms.uid() != in->uid) {
+      if((in->mode & S_ISUID) || (in->mode & S_ISGID)) {
```
Comment (ukernel) [inline src/client/Client.cc:5444]: this means that any user can modify mode of file with setuid/setgid. I don't think it's correct

Comment (ukernel) [reply src/client/Client.cc:5444]: the newest code still has this issue

**Pull request closed without merging**